### PR TITLE
Feature redis

### PR DESCRIPTION
**Pull request by HNDaG**

### CODE REVIEW

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/User.kt [hunk 1]
```diff
@@ -8,11 +10,13 @@ import org.springframework.data.mongodb.core.mapping.Document
 @Document(collection = COLLECTION_NAME)
 data class User(
     @Id
+    @JsonSerialize(using = ToStringSerializer::class)
     val id: ObjectId? = null,
     val username: String,
     val password: String,
     val email: String,
     val subscriptions: MutableSet<ObjectId> = mutableSetOf(),
+    @JsonSerialize(using = ToStringSerializer::class)
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/User.kt]: why do you need the annotation?

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/redis/config/RedisConfig.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+class RedisConfig {
+
+    @Value("\${spring.data.redis.host}")
+    lateinit var host: String
+
+    @Value("\${spring.data.redis.port}")
+    lateinit var port: String
+
+    @Bean
+    fun reactiveRedisConnectionFactory(): ReactiveRedisConnectionFactory {
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/redis/config/RedisConfig.kt]:
```suggestion

    @Bean
    fun reactiveRedisConnectionFactory(
        @Value("\${spring.data.redis.host}") host: String,
        @Value("\${spring.data.redis.port}") port: String,    
    ): ReactiveRedisConnectionFactory {
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/redis/repositiry/UserRedisRepositoryImpl.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+@Repository
+class UserRedisRepositoryImpl(
+    private val redisTemplate: ReactiveRedisTemplate<String, User>
+) : UserRedisRepository {
+
+    @Value("\${redis.ttl.minutes}")
+    private lateinit var redisTtlMinutes: String
+
+    @Value("\${redis.key.user-username-prefix}")
+    private lateinit var prefix: String
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/redis/repositiry/UserRedisRepositoryImpl.kt]: field injection. Avoid it

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/redis/repositiry/UserRedisRepositoryImpl.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/redis/repositiry/UserRedisRepositoryImpl.kt]:
```suggestion

```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/redis/repositiry/UserRedisRepositoryImpl.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        return redisTemplate.opsForValue().get(prefix + username)
+    }
+
+    override fun save(user: User): Mono<User> {
+        val key = user.username
+        return user.id?.let {
+            redisTemplate.opsForValue().set(
+                prefix + key, user, Duration.ofMinutes(redisTtlMinutes.toLong())
+            ).thenReturn(user)
+        } ?: Mono.empty()
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/redis/repositiry/UserRedisRepositoryImpl.kt]:
```suggestion
        Mono.just(user)
            .filter { user.id != null }
            .flatMap { 
                redisTemplate.opsForValue().set(
                        prefix + user.username, 
                        user,
                        Duration.ofMinutes(redisTtlMinutes.toLong())
                ).thenReturn(user)
           }
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt [hunk 6]
```diff
@@ -98,6 +107,7 @@ class UserServiceImpl(
             .flatMap { (user, publisher) ->
                 user.subscriptions.add(publisher.id!!)
                 userRepository.save(user)
+                redisUserRepository.update(user)
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt]:
```kotlin
                userRepository.save(user)
                redisUserRepository.update(user)
```

As I understand, the both invocations are reactive. 
So, with your changes `userRepository.save(user)` is not subscribed. I bet IDEA warn you about this

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt [hunk 7]
```diff
@@ -50,7 +52,10 @@ class PublisherServiceImpl(
     override fun deletePublisher(publisherName: String): Mono<Unit> {
         return publisherRepository.findByPublisherName(publisherName)
             .flatMap { publisher -> publisher.id?.let { userRepository.findByPublisherId(it) } ?: Mono.empty() }
-            .flatMap { user -> userRepository.save(user.copy(publisherId = null)) }
+            .flatMap { user ->
+                userRepository.save(user.copy(publisherId = null))
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt:57]: image

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt [hunk 8]
```diff
@@ -21,25 +21,33 @@ import reactor.kotlin.core.util.function.component2
...
-    }
-
-    override fun getUserById(id: String): Mono<User> {
-        return userRepository.findById(ObjectId(id))
-            .switchIfEmpty { NotFoundException("User not found with given id = $id").toMono() }
+        return redisUserRepository.findByUsername(username)
+            .switchIfEmpty {
+                userRepository.findByUsername(username)
+                    .flatMap { user ->
+                        redisUserRepository.save(user).thenReturn(user)
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt]: Why do you need `.thenReturn(user)`?

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt [hunk 9]
```diff
@@ -21,25 +21,33 @@ import reactor.kotlin.core.util.function.component2
...
+                    .switchIfEmpty(NotFoundException(message = "User not found with username = $username").toMono())
+            }
     }
 
     override fun createUser(user: User): Mono<User> {
-        return userRepository.findByUsername(user.username)
+        return redisUserRepository.findByUsername(user.username)
+            .switchIfEmpty {
+                userRepository.findByUsername(user.username)
+            }
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt]:
```kotlin
redisUserRepository.findByUsername(user.username)
            .switchIfEmpty {
                userRepository.findByUsername(user.username)
            }
```

This code repeated multiple times. Maybe it's time to extract a private function?

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt [hunk 10]
```diff
@@ -21,25 +21,33 @@ import reactor.kotlin.core.util.function.component2
...
             .handle<User> { _, sync ->
                 sync.error(DuplicateException("User with this username already exists"))
             }
-            .switchIfEmpty(userRepository.save(user))
+            .switchIfEmpty(userRepository.save(user)
+                .flatMap { savedUser ->
+                redisUserRepository.save(savedUser).thenReturn(savedUser)
+            })
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt]:
```suggestion
            .switchIfEmpty { 
                userRepository.save(user)
                    .flatMap { savedUser -> redisUserRepository.save(savedUser) }
            }
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt [hunk 11]
```diff
@@ -51,44 +59,45 @@ class UserServiceImpl(
...
-                if (deletedCount == 0L) {
-                    sync.error(NotFoundException("User with ID - $id not found"))
-                }
-            }
-    }
-
     override fun deleteUserByUsername(username: String): Mono<Unit> {
         return userRepository.deleteUserByUsername(username)
-            .handle { deletedCount, sync ->
+            .handle<Unit?> { deletedCount, sync ->
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt]:
```suggestion
            .handle<Unit> { deletedCount, sync ->
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt [hunk 12]
```diff
@@ -51,44 +59,45 @@ class UserServiceImpl(
...
-        return userRepository.findById(ObjectId(id))
-            .switchIfEmpty { NotFoundException("User not found with given id = $id").toMono() }
-            .flatMap { userRepository.save(user.copy(id = ObjectId(id))) }
+    override fun updateUser(oldUsername: String, user: User): Mono<User> {
+        return userRepository.findByUsername(oldUsername)
+            .switchIfEmpty { NotFoundException("User not found with username = $oldUsername").toMono() }
+            .flatMap {
+                userRepository.save(user.copy(id = it.id))
+                    .flatMap { savedUser ->
+                        redisUserRepository.update(savedUser).thenReturn(savedUser)
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt]:
```suggestion
                        redisUserRepository.update(savedUser)
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt [hunk 13]
```diff
@@ -51,44 +59,45 @@ class UserServiceImpl(
...
+                        userRepository.findByUsername(username)
+                    }
+                    .filter { user -> user.publisherId == null }
                     .switchIfEmpty { DuplicateException("User is already a publisher").toMono() }
                     .zipWhen { publisherRepository.save(publisher) }
                     .flatMap { (user, newPublisher) ->
                         val userAsPublisher = user.copy(publisherId = newPublisher.id)
-                        userRepository.save(userAsPublisher).thenReturn(newPublisher)
+                        userRepository.save(userAsPublisher).then(redisUserRepository.update(userAsPublisher))
+                            .thenReturn(newPublisher)
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt:58]:
```suggestion
```

**Review by ajax-gnatiuk-s — CHANGES_REQUESTED** [review]

**Comment by ajax-gnatiuk-s** [comment]
@HNDaG ask somebody from your colleagues to review this PR

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt (hunk 9 above)
Comment (HNDaG) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt]: Fixed that in other way

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/User.kt (hunk 1 above)
Comment (HNDaG) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/entity/User.kt]: Fixed(Separate entity for Redis)

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt (hunk 13 above)
Comment (HNDaG) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt:58]: Here I need to return ** newPublisher**

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/mongo/MongoUserRepository.kt [hunk 14]
```diff
@@ -14,39 +18,29 @@ import reactor.core.publisher.Flux
...
     override fun deleteUserByUsername(username: String): Mono<Long> {
         val query = Query().addCriteria(Criteria.where("username").`is`(username))
-        return mongoTemplate.remove<User>(query).map { it.deletedCount }
+        return mongoTemplate.remove<MongoUser>(query).map { it.deletedCount }
     }
 
     override fun findByPublisherId(publisherId: ObjectId): Mono<User> {
         val query = Query().addCriteria(Criteria.where("publisherId").`is`(publisherId))
-        return mongoTemplate.findOne<User>(query)
+        return mongoTemplate.findOne<MongoUser>(query).map { it.toEntity()}
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/repository/mongo/MongoUserRepository.kt]:
```suggestion
        return mongoTemplate.findOne<MongoUser>(query).map { it.toEntity() }
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt [hunk 15]
```diff
@@ -19,72 +19,60 @@ import reactor.kotlin.core.util.function.component2
...
                 sync.error(DuplicateException("Publisher with the same PublisherName already exists"))
             }
             .switchIfEmpty {
                 userRepository.findByUsername(username)
-                    .filter { user -> user.publisherId != null }
+                    .filter { user -> user.publisherId == null }
                     .switchIfEmpty { DuplicateException("User is already a publisher").toMono() }
-                    .zipWhen { publisherRepository.save(publisher) }
-                    .flatMap { (user, newPublisher) ->
+                    .zipWhen { publisherRepository.save(publisher) }.flatMap { (user, newPublisher) ->
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt]:
```suggestion
                    .zipWhen { publisherRepository.save(publisher) }
                    .flatMap { (user, newPublisher) ->
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt [hunk 16]
```diff
@@ -19,72 +19,61 @@ import reactor.kotlin.core.util.function.component2
...
-            .switchIfEmpty(userRepository.save(user))
+            .switchIfEmpty {
+                userRepository.save(user)
+            }
     }
 
     override fun becamePublisher(
-        username: String,
-        publisher: Publisher
+        username: String, publisher: Publisher
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/UserServiceImpl.kt]:
```suggestion
        username: String, 
        publisher: Publisher
```

File: listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt [hunk 17]
```diff
@@ -50,19 +53,24 @@ class PublisherServiceImpl(
...
+                    .then(redisUserRepository.update(user.copy(publisherId = null))) // todo: fix this
```
Comment (ajax-gnatiuk-s) [inline listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt:58]: > // todo: fix this

Let's fix :)
Or remove the comment

Comment (HNDaG) [reply listening-platform/src/main/kotlin/com/nikitahohulia/listeningplatform/service/PublisherServiceImpl.kt:58]: So just remove? OK)

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**